### PR TITLE
[Static if] replace overload constraints with static if (mutation.d)

### PR DESCRIPTION
**Pull request by wilzbach**
Split up of #5145

### CODE REVIEW

File: std/algorithm/mutation.d [hunk 1]
```diff
@@ -2132,24 +2127,67 @@ if (isBidirectionalRange!Range
...
+    {
+        while (!r.empty)
+        {
+            swap(r.front, r.back);
+            r.popFront();
+            if (r.empty) break;
+            r.popBack();
+        }
+    }
+    else static if (isNarrowString!Range && !is(ElementType!Range == const) && !is(ElementType!Range == immutable))
```
Comment (wilzbach) [inline std/algorithm/mutation.d]: As suggested this puts the overloads into one function, s.t. documentation & error message is improved

File: std/algorithm/mutation.d [hunk 2]
```diff
@@ -1745,138 +1745,133 @@ Returns:
...
-        result.popBackExactly(delta);
-        pos += delta;
+        // leftover move
+        moveAll(src, tgt);
+        return result;
     }
-    // leftover move
-    moveAll(src, tgt);
-    return result;
+    else static assert(0, "Unknown SwapStrategy.");
```
Comment (JackStouffer) [inline std/algorithm/mutation.d]: Nitpick, change to `SwapStrategy.semistable is not supported`

**Comment by wilzbach** [comment]
> Looks good otherwise

Addressed the nit.

**Comment by MartinNowak** [comment]
This broke the nightly builds @wilzbach, see http://nightlies.dlang.org/dmd-master-2017-05-13/build.html.
Please either fixup or revert this PR.

I'm also not positive this is a good long-term direction in general, hope we can work on improved error messages for template constraints later this year.

**Comment by JackStouffer** [comment]
@MartinNowak The motivation is documentation simplification and not error messages.

**Comment by MartinNowak** [comment]
https://dlang.org/phobos/std_algorithm_mutation.html#copy
I see, but now it no longer mentions the constraints/required concepts at all.
Of course a private and undocumentd `areCopyCompatibleArrays` constraints is an anti-pattern as well.

**Comment by MartinNowak** [comment]
Reduced test-case:
```
import std.file, std.algorithm;

void main()
{
    string a, b;
    copy(a, b);
}
```
This actually seems to break "instantiation failure is not an error", for the copy overloadset, because the constraints are too general, now both imported copy functions are callable, but one already errors during semantic (think within the overload resolution).
Please have a look at preserving the existing constraints.

**Pull request merged**